### PR TITLE
LibriSpeech-100 GMM baseline

### PR DESCRIPTION
**Pull request by JackTemaki**
So, after some work, here is the PR for the first official GMM-Pipeline.

A lot of documentation is still missing, and the recognition settings are still debatable, but the training itself should be fine. @Marvin84 and I run this pipeline already in multiple small variations, and also @christophmluscher already did tests for 960h and will also add his Pipeline afterwards.

Please only comment on:
 - severely wrong parameters (for training, recognition could be discussed again later)
 - file naming/structure
 
 Everything else, e.g. naming, code format, documentation, can be discussed later as it has no effect on the experiment graph.

### CODE REVIEW

File: common/baselines/librispeech/ls100/gmm/baseline_args.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+                "normalization_options": None,
+                "without_samples": False,
+                "samples_options": samples_options,
+                "cepstrum_options": {
+                    "normalize": False,
+                    "outputs": 16,  # this is the actual output feature dimension
+                    "add_epsilon": not dc_detection,  # when there is no dc-detection we can have log(0) otherwise
+                    "epsilon": 1e-10,
+                },
+                "fft_options": None,
```
Comment (michelwi) [inline common/baselines/librispeech/ls100/gmm/baseline_args.py:67]: "fft_options" here is `None` and below at "energy" it is `{}`. Is this intended?

File: common/baselines/librispeech/ls100/gmm/baseline_args.py [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+                "fft_options": {},
+            }
+        },
+    }
+
+    return RasrInitArgs(
+        costa_args=costa_args,
+        am_args=am_args,
+        feature_extraction_args=feature_extraction_args,
+        default_mixture_scorer_args={"scale": 0.3},
```
Comment (michelwi) [inline common/baselines/librispeech/ls100/gmm/baseline_args.py:83]: does this set the emission model scale to 0.3? Did we ever use a value different from 1.0?

File: common/baselines/librispeech/ls100/gmm/baseline_args.py [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+    )
+
+
+def get_cart_args(
+    use_stress_marker: bool = False,
+    max_leaves: int = 12001,
+    min_obs: int = 1000,
+    hmm_states: int = 3,
+    feature_flow: str = "mfcc+deriv+norm",
+    add_unknown: bool = True,
```
Comment (michelwi) [inline common/baselines/librispeech/ls100/gmm/baseline_args.py:148]: should the default here maybe match the default in `common/baselines/librispeech/data.py` where in the lexicon unknown phoneme is added iff `not use_g2p_training` with `use_g2p_training=True` by default --> no unknown phoneme by default.
```suggestion
    add_unknown: bool = False,
```

File: common/baselines/librispeech/ls100/gmm/baseline_args.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+
+The parameters defined here are based on the past experience with training LibriSpeech-100h and 960h models,
+but no excessive tuning has been done.
+"""
+from i6_core.features.filterbank import filter_width_from_channels
+from i6_core import cart
+
+from i6_experiments.common.setups.rasr import gmm_system
+
+from i6_experiments.common.setups.rasr.util import *
```
Comment (michelwi) [inline common/baselines/librispeech/ls100/gmm/baseline_args.py:12]: sorry, I know you explicitly asked to not comment on this, but can we be explicit on which things are imported from where?

File: common/baselines/librispeech/ls100/gmm/baseline_config.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+from i6_experiments.common.baselines.librispeech.data import get_corpus_data_inputs
+
+from ...default_tools import RASR_BINARY_PATH
+
+
+def run_librispeech_100_common_baseline(
+    alias_prefix="baselines/librispeech/ls100/gmm/common_baseline",
+):
+
+    gs.ALIAS_AND_OUTPUT_SUBDIR = alias_prefix
```
Comment (michelwi) [inline common/baselines/librispeech/ls100/gmm/baseline_config.py:19]: maybe store the previous value of `gs.ALIAS_AND_OUTPUT_SUBDIR` in a local variable to restore that later.

File: common/baselines/librispeech/ls100/gmm/baseline_config.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    hybrid_init_args = baseline_args.get_init_args()
+    mono_args = baseline_args.get_monophone_args()
+    cart_args = baseline_args.get_cart_args()
```
Comment (michelwi) [inline common/baselines/librispeech/ls100/gmm/baseline_config.py:23]: now here is a wrong parameter I think: as you do not give any options to `get_cart_args()`, the returned questions will contain the unknown phoneme which is incompatible with the lexicon returned below using `use_g2p_training=True`

File: common/baselines/librispeech/ls100/gmm/baseline_config.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    steps.add_step("extract", hybrid_init_args.feature_extraction_args)
+    steps.add_step("mono", mono_args)
+    steps.add_step("cart", cart_args)
+    steps.add_step("tri", tri_args)
+    steps.add_step("vtln", vtln_args)
+    steps.add_step("sat", sat_args)
+    steps.add_step("vtln+sat", vtln_sat_args)
+    steps.add_step("output", final_output_args)
+
+    corpus_data = get_corpus_data_inputs("train-clean-100", use_g2p_training=True)
```
Comment (michelwi) [inline common/baselines/librispeech/ls100/gmm/baseline_config.py:46]: I think `use_g2p_training=True` is the default, so you can leave it out. If you want to include it to be more explicit about it, fine, but I think you should also/rather be explicit about the unstressed markers.

File: common/baselines/librispeech/ls100/gmm/baseline_args.py [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+    }
+
+    am_args = {
+        "state_tying": "monophone",
+        "states_per_phone": 3,
+        "state_repetitions": 1,
+        "across_word_model": True,
+        "early_recombination": False,
+        "tdp_scale": 1.0,
+        "tdp_transition": (3.0, 0.0, 30.0, 0.0),  # loop, forward, skip, exit
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply common/baselines/librispeech/ls100/gmm/baseline_args.py:33]: @christophmluscher @Marvin84?

File: common/baselines/librispeech/ls100/gmm/baseline_args.py (hunk 1 above)
Comment (JackTemaki) [reply common/baselines/librispeech/ls100/gmm/baseline_args.py:67]: No, will fix

Comment (JackTemaki) [reply common/baselines/librispeech/ls100/gmm/baseline_args.py:67]: Actually this is a little bit more complicated, I will do a larger refactoring on the ` i6_core` side.

File: common/baselines/librispeech/ls100/gmm/baseline_args.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (JackTemaki) [reply common/baselines/librispeech/ls100/gmm/baseline_args.py:33]: After personal discussion: "infinity" is fine.

File: common/baselines/librispeech/ls100/gmm/baseline_args.py (hunk 2 above)
Comment (JackTemaki) [reply common/baselines/librispeech/ls100/gmm/baseline_args.py]: Was not used anyway, removed.

**Review by Marvin84 — APPROVED** [review]
By taking a quick look, everything seems fine.

File: common/baselines/librispeech/ls100/gmm/baseline_args.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (michelwi) [reply common/baselines/librispeech/ls100/gmm/baseline_args.py:33]: @JackTemaki your commit message for https://github.com/rwth-i6/i6_experiments/commit/92fad0e4d08979781d5ee8fbacea4df08fcc09e0 said "tdp skip to infinity" but it did not contain the change. FYI

[the remark this answers is not in the document]
Comment (JackTemaki) [reply common/baselines/librispeech/ls100/gmm/baseline_args.py:33]: @michelwi good catch

File: common/baselines/librispeech/default_tools.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+"""
+List of default tools and software to be defined as default independent from hashing
+by setting one explicit hash.
+
+In order to use different software paths without hash changes, just use the same explicit hash string as given here.
+
+If you want a stronger guarantee that you get the intended results, please consider using the explicit software
+version listed here. Nevertheless, the most recent "head" should be safe to be used as well
+
+"""
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply common/baselines/librispeech/default_tools.py:10]: SCTK scorer will follow in another PR.

**Comment by JackTemaki** [comment]
@christophmluscher please be aware that I also added `add_features_outputs` to the MFCC flow parameters. This was marked by Eugen as "should be done for new pipelines", so do not forget to put this for the 960h as well.

**Review by michelwi — APPROVED** [review]

File: common/baselines/librispeech/ls100/gmm/baseline_args.py [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+                "without_samples": False,
+                "samples_options": samples_options,
+                "cepstrum_options": {
+                    "normalize": False,
+                    "outputs": 16,  # this is the actual output feature dimension
+                    "add_epsilon": not dc_detection,  # when there is no dc-detection we can have log(0) otherwise
+                    "epsilon": 1e-10,
+                },
+                "fft_options": None,
+                "add_feature_output": True,
```
Comment (Atticus1806) [inline common/baselines/librispeech/ls100/gmm/baseline_args.py:66]: In `i6_core` there is an additional s here. Not sure if this is wrong here or in core, but I am inclined to think here. 
```suggestion
                "add_features_output": True,
```

Comment (Atticus1806) [reply common/baselines/librispeech/ls100/gmm/baseline_args.py:66]: or is this also included in the refactoring on core side stated above?

File: common/baselines/librispeech/ls100/gmm/baseline_args.py [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+    sat_training_args = {
+        "name": "sat",
+        "mixtures": "estimate_mixtures_sdm.tri",
+        "alignment": "train_tri",
+        "feature_cache": "mfcc",
+        "feature_flow_key": "mfcc+context+lda",
+        "cache_regex": "^mfcc.*$",
+        "splits": 10,
+        "accs_per_split": 2,
+        "align_keep_values": {7: 20},
```
Comment (christophmluscher) [inline common/baselines/librispeech/ls100/gmm/baseline_args.py:318]: not necessary

**Review by christophmluscher — APPROVED** [review]

**Pull request merged**